### PR TITLE
Fixup precommit and update Ruff

### LINKED ISSUE #16852: Consider formatting markdown according to pre-commit expectations
cc @seanbudd 

### Is your feature request related to a problem? Please describe.
With the current pre-commit config, commits including changes.md will perform a major rewrite of the file due to trailing whitespace

### Describe the solution you'd like
Format the docs according to pre-commit expectations.

### Describe alternatives you've considered
1. Exclude docs from pre-commit
2. Disable pre-commit on merge. It seems almost impossible to perform a merge with conflicts without pre-commit kicking in

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:

Follow up to #16767, #16803
Preparation for #16852 

### Summary of the issue:

Pre-commit touches several files it shouldn't touch, e.g. symbol dictionaries, translated markdown.

### Description of user facing changes

None

### Description of development approach

- Update Ruff to 0.5.2 in requirements
- Add a hook for `scons source`
- Update `sconstruct` to automatically use the max number of cores when passing `-j 0`, which can be used with pre-commit to have optimal performance
- Run pre-commit autoupdate

Note, the following script can be used to perform the actual vfixup that was initially part of this pr, but @seanbudd requested it to be separated.

```batchfile
pre-commit autoupdate
# lint scons files
ruff check --fix
ruff check --add-noqa

# add trailing commas
pre-commit run --all-files add-trailing-comma

# format and re lint
ruff format
ruff check --fix
pre-commit run --all-files
```

### Testing strategy:
- [ ] Test that python file changes will runn checkpot and unit tests (it didn't before)

### Known issues with pull request:
None known

### Code Review Checklist:

* [x]  Documentation:
      
  * Change log entry
  * User Documentation
  * Developer / Technical Documentation
  * Context sensitive help for GUI changes

* [x]  Testing:
      
  * Unit tests
  * System (end to end) tests
  * Manual testing

* [x]  UX of all users considered:
      
  * Speech
  * Braille
  * Low Vision
  * Different web browsers
  * Localization in other languages / culture than English

* [x]  API is compatible with existing add-ons.

* [x]  Security precautions taken.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -35,13 +39,13 @@ repos:
       entry: cmd.exe /c "scons checkPot -j 4"
       language: system
       pass_filenames: false
-      types: [python, pofile]
```
Comment (LeonarddeR) [inline .pre-commit-config.yaml:38]: This would never run when python files would be changed. Note that you need `types_or` to handle multiple types, using `types` expects a file to match all the types.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -7,8 +9,10 @@ repos:
     - id: check-ast
     - id: check-case-conflict
     - id: check-merge-conflict
-    - id: end-of-file-fixer
     - id: trailing-whitespace
+      types_or: [python, c, c++, batch, markdown]
+    - id: end-of-file-fixer
```
Comment (LeonarddeR) [inline .pre-commit-config.yaml:14]: Ive seen cases where applying trim trailing whitespace will cause end of file to fail again, so that's why I swapped them.

Comment (seanbudd) [reply .pre-commit-config.yaml:14]: these don't appear swapped?

**Review by seanbudd — CHANGES_REQUESTED** [review]
Do you mind doing the first commit (settings fixup) as a separate PR to the actual fixups (i.e. running this on all files).
It will make reviewing easier.
I'd also like NV Access to be the one to perform the actual fixes, so we don't have to manually review all changes in as much thorough detail, and it is up to date at a point where we can review it

**Comment by seanbudd** [comment]
This diff generally looks fine to me

File: .pre-commit-config.yaml (hunk 2 above)
Comment (LeonarddeR) [reply .pre-commit-config.yaml:14]: `trailing-whitespace` is definitely above `end-of-file-fixer` now. It was the other way around before this pr.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -35,13 +39,13 @@ repos:
...
       pass_filenames: false
-      types: [python, pofile]
+      types: [python]
     - id: unitTest
       name: unit tests
       entry: ./rununittests.bat
       language: script
       pass_filenames: false
-      types: [python, c, c++, batch]
+      types_or: [python, c, c++, batch]
```
Comment (LeonarddeR) [inline .pre-commit-config.yaml:48]: I wonder whether it makes sense to run the tests when c/c++ files have changed. It is necessary to run scons source first when that's the case. May be we should add that? What do you think @seanbudd?

Comment (seanbudd) [reply .pre-commit-config.yaml:48]: Good question - I'm not sure. I think we need to run tests when certain dependencies change - e.g. the braille tables test when liblouis changes. This probably requires a scons source

Comment (seanbudd) [reply .pre-commit-config.yaml:48]: I think adding scons source to pre-commit is a good idea for c++ changes, to ensure NVDA builds successfully

File: user_docs/en/changes.md [hunk 4]
```diff
@@ -26,6 +26,8 @@ Please refer to [the developer guide](https://www.nvaccess.org/files/nvda/docume
 
 * Added a `.editorconfig` file to NVDA's repository in order for several IDEs to pick up basic NVDA code style rules by default. (#16795, @LeonarddeR)
 * It is now possible to redirect objects retrieved from on-screen coordinates, by using the `NVDAObject.objectFromPointRedirect` method. (#16788, @Emil-18)
+* Running SCons with the parameter `-j0` will automatically pick the maximum number of available CPU cores. (#16868 , @LeonarddeR)
+* Updated Ruff to version 0.5.2. (#16868 , @LeonarddeR)
```
Comment (Qchristensen) [inline user_docs/en/changes.md:30]:
```suggestion
* Updated Ruff to version 0.5.2. (#16868, @LeonarddeR)
```

File: user_docs/en/changes.md [hunk 5]
```diff
@@ -26,6 +26,8 @@ Please refer to [the developer guide](https://www.nvaccess.org/files/nvda/docume
 
 * Added a `.editorconfig` file to NVDA's repository in order for several IDEs to pick up basic NVDA code style rules by default. (#16795, @LeonarddeR)
 * It is now possible to redirect objects retrieved from on-screen coordinates, by using the `NVDAObject.objectFromPointRedirect` method. (#16788, @Emil-18)
+* Running SCons with the parameter `-j0` will automatically pick the maximum number of available CPU cores. (#16868 , @LeonarddeR)
```
Comment (Qchristensen) [inline user_docs/en/changes.md:29]:
```suggestion
* Running SCons with the parameter `-j0` will automatically pick the maximum number of available CPU cores. (#16868, @LeonarddeR)
```

**Review by Qchristensen — CHANGES_REQUESTED** [review]
UserGuide looks fine, just removed the spaces between the issue number and the comma

**Comment by LeonarddeR** [comment]
thanks @Qchristensen

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -30,18 +34,24 @@ repos:
 
 - repo: local
   hooks:
+    - id: scons-source
+      name: running build system
```
Comment (seanbudd) [inline .pre-commit-config.yaml:38]:
```suggestion
      name: build c/c++ files
```

File: sconstruct [hunk 7]
```diff
@@ -162,11 +162,15 @@ env.Decider("MD5-timestamp")
...
+if numJobs == 0:
+	numJobs = numCores
+	env.SetOption('num_jobs', numJobs)
 if numJobs < numCores:
 	print(
 		f"Warning: Building with {numJobs} concurrent job{'s' if numJobs != 1 else ''} "
 		f"while {numCores} CPU threads are available. "
-		f"Running SCONS with the parameter '-j{numCores}' may lead to a faster build."
+		f"Running SCONS with the parameter '-j{numCores}' may lead to a faster build. "
+		"Running SCONS with parameter '-j0' will automatically pick all available cores."
```
Comment (seanbudd) [inline sconstruct:173]:
```suggestion
		"Running SCONS with parameter '-j0' will automatically pick all available cores. "
```

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -30,18 +34,24 @@ repos:
...
     - id: unitTest
       name: unit tests
       entry: ./rununittests.bat
       language: script
       pass_filenames: false
-      types: [python, c, c++, batch]
+      types_or: [python, c, c++, batch]
     - id: licenseCheck
       name: Check license compatibility of pip dependencies
       files: requirements.txt
```
Comment (seanbudd) [inline .pre-commit-config.yaml:57]: can you add `./runlicensecheck.bat` to this?

Comment (LeonarddeR) [reply .pre-commit-config.yaml:57]: @coderabbitai could you provide a suggestion on how to add `runlicensecheck.bat` to this, i.e. the `licenseCheck` should run when either `requirements.txt` or `runlicensecheck.bat` changed?

Comment (LeonarddeR) [reply .pre-commit-config.yaml:57]: Provide a suggestion in a suggestion block please

Comment (LeonarddeR) [reply .pre-commit-config.yaml:57]: @coderabbitai sorry I was talking to you, please provide a suggestion block so I can commit it

Comment (LeonarddeR) [reply .pre-commit-config.yaml:57]: @coderabbitai This is not what I meant
Please take note of this:

```suggestion
      files: ^(requirements\.txt|runlicensecheck\.bat)$
```

Comment (LeonarddeR) [reply .pre-commit-config.yaml:57]: @coderabbitai No, you are still mistaken. Please analyse comment https://github.com/nvaccess/nvda/pull/16868#discussion_r1683831119 thoroughly. You should provide a fenced code block that starts with

    ```suggestion
          files: ^(requirements\.txt|runlicensecheck\.bat)$
    ```

Comment (LeonarddeR) [reply .pre-commit-config.yaml:57]: @coderabbitai when we ask you for code suggestions, we want you to always use this format. Could you provide me info on how to update `.coderabbit.yml` so you won't forget this in future?

File: .pre-commit-config.yaml
Comment (LeonarddeR) [inline .pre-commit-config.yaml:1]: @coderabbitai please review this. If there is an `exclude` or `files` entry, explain what the regex does in its context

Comment (LeonarddeR) [reply .pre-commit-config.yaml:1]: @coderabbitai you missed the entry for the checkpot hook

Comment (seanbudd) [reply .pre-commit-config.yaml:1]: can you provide a code suggestion to fix this?

File: .coderabbit.yml [hunk 9]
```diff
@@ -26,6 +26,10 @@ reviews:
       This project uses tabs instead of spaces, do not suggest usage of spaces over tabs.
       Are there any 'red flags' in this code that might warrant closer investigation from a security standpoint?
       Explain what makes them suspicious.
+      When providing in line code suggestions, always use the following format:
+      ```suggestion
+      <code changes>
```
Comment (seanbudd) [inline .coderabbit.yml:31]: I'm not sure this is going to always be appropriate. I think suggestions are the default for reviews when it is confident, and you can always ask in comment replies

File: venvUtils/ensureVenv.py [hunk 10]
```diff
@@ -179,7 +179,7 @@ def ensureVenvAndRequirements():
 			"This means that transitive dependencies can get out of sync "
 			"with those used in automated builds. "
 			"Would you like to continue recreating the environment?",
-			default=True,
+			default=False,
```
Comment (seanbudd) [inline venvUtils/ensureVenv.py:182]: why is this changed?

File: .coderabbit.yml (hunk 9 above)
Comment (LeonarddeR) [reply .coderabbit.yml:31]: As you have seen, it is pretty hard to convince code rabbit of the proper form when asking for suggestions:

```suggestion
      When someone asks for code suggestions in an inline comment, always use the following format:
      ```suggestion
      <code changes>
```

File: venvUtils/ensureVenv.py (hunk 10 above)
Comment (LeonarddeR) [reply venvUtils/ensureVenv.py:182]: Ah, I only explained that in the commit summary.
When pre-commit runs something in the NVDA venv, it does so non-interactively. When requirements changed, the True value will effectively mean that pre-commit will recreate the venv. That is really overkill in many situations. Ive also encountered cases myself where I ran this non-interactively, causing my venv to be discarded.
I think it still makes sense to encourage people to recreate the venv especially if they don't know what they're doing, but non-interactive automation should impose the smallest possible impact on the venv IMO.

File: .pre-commit-config.yaml (hunk 8 above)
Comment (seanbudd) [reply .pre-commit-config.yaml]: I think the issue here is that this line was initially outside of the diff, so suggestions cannot be applied, even if the view is more  useful

File: .coderabbit.yml (hunk 9 above)
Comment (seanbudd) [reply .coderabbit.yml:31]: As you have seen, it is pretty hard to convince code rabbit of the proper form when asking for suggestions:

```suggestion
      When providing code suggestions, particularly when requested, ensure GitHub's suggestion format is used, i.e.:
      ```suggestion
      <code changes>
```

**Review by seanbudd — APPROVED** [review]

**Comment by seanbudd** [comment]
I'll be merging this tomorrow morning AEST

**Comment by LeonarddeR** [comment]
I noticed that the -0 trick actually doesn't work, since as soon as you override a command line option, you can't set it with SetOption. See https://github.com/SCons/scons/issues/3937
